### PR TITLE
Move most video playback into global queue - Part 1 of conversion

### LINKED ISSUE #1135: Video overlay loads before overhang is gone when using global video player
## Software Versions

- Jellyfin Server Version:  demo stable
- Roku Client Version:  latest unstable

## Describe the bug
Using the shuffle buttons causes the video overlay (or menu?) to load before the overhang is gone. This causes the two clocks to be shown overlapping for a brief time

## How To Reproduce

1. open demo server
2. go to tv library
3. open only show
4. open only season
5. hit "Shuffle" button
6. observe overhang area specifically the top right of the screen

## Expected behavior
overhang disappears immediately before the video overlay joins the party

## Connection Information

- Is server local or remote? reomte
- Is server connection HTTP or HTTPS? HTTPS

### PR DESCRIPTION
**Pull request by 1hitsong**
Changes most video players from various locations to use new global queue playback functionality - like playlists already use.

All functionality should be the same as the old method. Meaning everything from selecting alternative video & audio tracks to playing local trailers and movie extras should work as expected.

This is only part 1 of this conversion and the following video types have not been converted yet:

1. Deep Link
2. Search items
3. TV Channel finished state
4. TV Channel error retry

As part of this change, there is existing code that was either depreciated or moved. A running list of these changed are here: https://github.com/jellyfin/jellyfin-roku/issues/1132 

Linked here is a list of tests that were performed and will need to be verified/reviewed as working properly.
[Roku global video test list.txt](https://github.com/jellyfin/jellyfin-roku/files/11413708/Roku.global.video.test.list.txt)

### CODE REVIEW

File: components/ItemGrid/LoadVideoContentTask.brs [hunk 1]
```diff
@@ -10,30 +10,45 @@ import "pkg:/source/utils/deviceCapabilities.brs"
...
+    m.top.isIntro = false
+
+    ' Only show preroll once per queue
+    if m.global.queueManager.callFunc("prerollActive")
+        ' Prerolls not allowed if we're resuming video
+        if m.global.queueManager.callFunc("getCurrentItem").startingPoint = 0
+            preRoll = GetIntroVideos(m.top.itemId)
+            if isValid(preRoll) and preRoll.TotalRecordCount > 0 and isValid(preRoll.items[0])
+                ' Bypass joke pre-roll
+                if lcase(preRoll.items[0].name) <> "rick roll'd"
```
Comment (cewert) [inline components/ItemGrid/LoadVideoContentTask.brs]: Joke pre-rolls? Rick roll'd? 😆 Why are we skipping these? I don't use prerolls so I'm probably missing something

File: components/ItemGrid/LoadVideoContentTask.brs [hunk 2]
```diff
@@ -60,42 +76,24 @@ sub LoadItems_AddVideoContent(video as object, mediaSourceId as dynamic, audio_s
...
+    user = AboutMe()
+    if user.Configuration.EnableNextEpisodeAutoPlay
+        if LCase(m.top.itemType) = "episode"
+            addNextEpisodesToQueue(video.showID)
         end if
     end if
 
-    ' For phase 1 of playlist support, we don't support intros yet
-    showIntro = false
+    playbackPosition = 0!
```
Comment (cewert) [inline components/ItemGrid/LoadVideoContentTask.brs:87]: Why force `playbackPosition` to be a float?

File: source/Main.brs [hunk 3]
```diff
@@ -603,6 +620,44 @@ sub Main (args as dynamic) as void
...
+                        m.global.queueManager.callFunc("push", selectedItem[0])
+                        m.global.queueManager.callFunc("playQueue")
+                    else if msg.getRoSGNode().returnData.indexselected = 2
+                        ' User chose Go to Series
+                        CreateSeriesDetailsGroup(selectedItem[0].json.SeriesId)
+                    else if msg.getRoSGNode().returnData.indexselected = 3
+                        ' User chose Go to season
+                        CreateSeasonDetailsGroupByID(selectedItem[0].json.SeriesId, selectedItem[0].json.seasonID)
+                    else if msg.getRoSGNode().returnData.indexselected = 4
+                        ' User chose Go to season
```
Comment (cewert) [inline source/Main.brs]:
```suggestion
                        ' User chose Go to episode
```

File: components/ItemGrid/LoadVideoContentTask.brs (hunk 1 above)
Comment (1hitsong) [reply components/ItemGrid/LoadVideoContentTask.brs]: Normally I'd be against such a blanket condition such as this; however, the intro plugin uses the rick roll music video as the fallback video if any error occurs. It's literally the fallback within the catch clause! So if any error occurs, we don't receive the error, we receive the 5 minute music video as if everything loaded properly 👎 

This is why I skipped it. I'm treating it as our identifier that an error occurred and the real/requested intro failed to load.

File: components/ItemGrid/LoadVideoContentTask.brs (hunk 2 above)
Comment (1hitsong) [reply components/ItemGrid/LoadVideoContentTask.brs:87]: StartingPoint on line 91 is in ticks, which can overflow an int. So we need it as a float until we convert ticks to seconds on line 95.

File: components/ItemGrid/LoadVideoContentTask.brs (hunk 1 above)
Comment (cewert) [reply components/ItemGrid/LoadVideoContentTask.brs]: That's as ridiculous as it is funny 😆 We should probably add this info as a comment. Something like "when intro plugin fails to load intro it returns a rick roll video"

**Review by cewert — CHANGES_REQUESTED** [review]
Everything seems to work well in my tests as far as playback and functionality (You even fixed #1135 in this PR). The new global video player is using different dialog popups for the up and down arrow functions while a video is playing. Style wise I'm fine either way but the new dialog is a big downgrade for UX when selecting subtitles (down arrow while video is playing):
- The menu options aren't selected by default vs old menu they were
- The user needs to select his language and also select the "okay" button when done to save his setting and close the dialog vs old menu saving and closing when the user selects his language (use episode 2 on the demo server to see how annoying this can be)
- If there are a lot of languages listed in the new menu, you have to "scroll" first on the right side before being able to see the buttons on the left side vs the old menu that loops the whole list when hitting down (use episode 2 on the demo server to see this in action)

**Comment by 1hitsong** [comment]
@cewert Thanks for the testing! Can you elaborate a bit on what this item means: "The menu options aren't selected by default vs old menu they were"

Edit: I think I may understand now. Do you mean the list of choices are not in cursor focus, the scrollbar is? I thought you meant the radio button wasn't in a selected state.

**Comment by cewert** [comment]
> Can you elaborate a bit on what this item means: "The menu options aren't selected by default vs old menu they were"

The new dialog basically has two "columns" the list of languages and the "scroll bar" vs the old menu with just one column with the list of languages. The new menu has the "scroll bar" selected by default adding an extra click.

So even when the list of languages is small - say one item. The user still runs into the first two issues listed and needs to press `left, down, okay, down, okay` vs the old menu which would be `down, okay`.

File: source/Main.brs [hunk 4]
```diff
@@ -603,6 +620,44 @@ sub Main (args as dynamic) as void
                     end if
                 end if
             end if
+        else if isNodeEvent(msg, "dataReturned")
+            if isValid(msg.getRoSGNode()) and isValid(msg.getRoSGNode().returnData)
```
Comment (cewert) [inline source/Main.brs]:
```suggestion
			msgNode = msg.getRoSGNode()
            if isValid(msgNode) and isValid(msgNode.returnData)
```

I think it makes sense to save the msg node rather retrieving it every time. Should be faster and makes the code a little easier to read.

**Comment by 1hitsong** [comment]
I've updated how the subtitle section popup operates.

When a user presses OK to select a subtitle, the cursor now jumps to the OK button, allowing the user to simply press OK once more to confirm their selection. I feel this confirmation is important as it gives context to what is happening next - especially if we have to reload the video because the subtitles are encoded. This also prevents reload if the user accidentally selected the wrong subtitles and needs to change their selection.

If there are so many subtitles that it goes off the popup, the option list now scrolls with the user's cursor. No more scrollbar, it follows the cursor.

The option list has initial focus.

File: components/manager/QueueManager.brs [hunk 5]
```diff
@@ -1,10 +1,18 @@
...
+import "pkg:/source/api/Image.brs"
+import "pkg:/source/utils/deviceCapabilities.brs"
 
 sub init()
+    m.hold = []
     m.queue = []
     m.originalQueue = []
     m.queueTypes = []
+    ' Preroll videos only play if user has cinema mode setting enabled
+    m.prerollActive = (get_user_setting("playback.cinemamode") = "true")
```
Comment (cewert) [inline components/manager/QueueManager.brs]:
```suggestion
    m.isPrerollActive = (get_user_setting("playback.cinemamode") = "true")
```

File: components/manager/QueueManager.brs [hunk 6]
```diff
@@ -107,6 +144,16 @@ sub pop()
     m.queueTypes.pop()
 end sub
 
+' Return prerollActive status
+function prerollActive()
```
Comment (cewert) [inline components/manager/QueueManager.brs]:
```suggestion
function isPrerollActive() as boolean
```

File: components/manager/QueueManager.brs [hunk 7]
```diff
@@ -107,6 +144,16 @@ sub pop()
     m.queueTypes.pop()
 end sub
 
+' Return isPrerollActive status
+function isPrerollActive() as boolean
+    return m.isPrerollActive
+end function
+
+' Set prerollActive status
+sub setPrerollStatus(newStatus)
```
Comment (cewert) [inline components/manager/QueueManager.brs]:
```suggestion
sub setPrerollStatus(newStatus as boolean)
```

I missed this one before sorry for so many code reviews

**Review by cewert — COMMENTED** [review]
The new dialog is much better now 👍  Only one extra click to hit "OK" but you made a good argument for it so that's fine by me. Not sure if these are all in the scope of this PR but here's my notes from testing:

- List of languages doesn't loop around. When there are less than like 5 languages to choose from this is a moot point. The only use case I can think of would be someone who speaks more than 1 language but has a preferred language. They may want to scroll the entire list of options before selecting their language and since the focus jumps to the OK button after the user selects a language the looping shouldn't interfere with anything.
- Now that the scroll bar is hidden, the list of languages doesn't line up with the header and OK button. This one is completely visual and doesn't affect functionality. It just looks "off"
- When you have your system subtitle setting on "replay only", opening the subtitle menu and selecting "none" then "ok" changes the system setting to "off". When or how we should be changing the system setting in general I'm not sure the best route forward but I don't think we should change the system setting when the user selects "none" when the setting was already set to "none". Hopefully that makes sense 😄 Edit: [made a new issue regarding this](https://github.com/jellyfin/jellyfin-roku/issues/1259)
- If the subtitle popup or stream info popup is open when the video finishes playing and kicks you back to the screen you came from, the popup remains open.

The last two may also be happening with the old popups. I didn't check.

**Comment by 1hitsong** [comment]
> When you have your system subtitle setting on "replay only", opening the subtitle menu and selecting "none" then "ok" changes the system setting to "off". When or how we should be changing the system setting in general I'm not sure the best route forward but I don't think we should change the system setting when the user selects "none" when the setting was already set to "none". Hopefully that makes sense 😄

This is how the system works now, not a regression nor a change. If we want to change how that works, we should do it outside of this PR and open a new issue ticket for it.

> If the subtitle popup or stream info popup is open when the video finishes playing and kicks you back to the screen you came from, the popup remains open.

The is a change from how the system works now. I'll take a look at it along with the first 2.

**Comment by 1hitsong** [comment]
> List of languages doesn't loop around. When there are less than like 5 languages to choose from this is a moot point. The only use case I can think of would be someone who speaks more than 1 language but has a preferred language. They may want to scroll the entire list of options before selecting their language and since the focus jumps to the OK button after the user selects a language the looping shouldn't interfere with anything.

We may have to revisit this one. I worked on it for several hours and didn't get anything working.

Other than the 1 existing item, I'm done with the others you mentioned.

**Comment by cewert** [comment]
Making the scroll bar visible but not selectable and having it move with the list is clever and looks slick. It's also functional in that it helps the user know how far down the list they are. Very nice!

A couple notes. One is easily fixable and the other one is an edge case that could wait till later

- On the subtitle popup, hitting "right" takes the user down to the "ok" button. (this is the easy one)
- On the subtitle popup, hitting "left" then "right" selects the scroll bar. Once the scroll bar is "selectable" it stays that way until the popup is closed

File: components/RadioDialog.brs [hunk 8]
```diff
@@ -1,35 +1,120 @@
...
+    if key = "up"
+        ' By default UP from the OK button is the scrollbar
+        ' Instead, move the user to the option list
+        if not m.radioOptions.isinFocusChain()
+            m.radioOptions.setFocus(true)
+            return true
+        end if
+    end if
+
+    return false
```
Comment (cewert) [inline components/RadioDialog.brs:137]:
```suggestion
    if key = "right"
        return true
    end if
    
    return false
```

This fixes the "right" button selecting the ok button while highlighting a language but may not be best practice? I don't mess around in `onKeyEvent()` too much 😄

**Review by cewert — APPROVED** [review]
Latest commit fixed both of the new issues 👍

**Pull request merged**